### PR TITLE
Verify member is in scope before trying to replace it

### PR DESCRIPTION
**Pull request by sharwell**
This corrects the largest low-hanging fruit for Fix All operations involving Simplify Type Name. Prior to creating a temporary workspace with an expression replaced (used to provide the validity of a replacement), we do a quick check using the C# speculative semantic model to ensure that the unqualified name at least has a possibility of resolving to the same member.

The time to execute a Fix All in Project on the CodeAnalysis project was previously dominated by examination of **WellKnownMembers.cs**. Each of the enum members referenced in the static constructor cannot be simplified, and this check will fail for each of those cases. This change is good for a 100+ second _wall clock_ time reduction in the Fix All in Project performance for this project.

## Ask Mode

**Customer scenario**

Attempt to use Fix All in a large project/solution for one of the name simplification diagnostics.

**Bugs this fixes:**

None that I could find, which surprised me.

**Workarounds, if any**

Simplify names one at a time, and/or by hand.

**Risk**

I would consider this a medium risk, with high reward. The @dotnet/roslyn-compiler team should verify that this code is either not used for compilation correctness, or is truly correct in all cases. Unit tests revealed several edge cases which have been accommodated to ensure behavior is unchanged, except for a small edge case described in #19368 which is knowingly deferred.

**Performance impact**

This change dramatically improves performance of the Simplify Type Name analyzer operating on C# code. It may also improve the simplification process in the formatter.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

The previous behavior was correct, but inefficient. PerfView was used to identify the majority offending code.

**How was the bug found?**

It became a clear problem during my attempts to scale analyzer features to Roslyn.sln.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 1]
```diff
@@ -812,6 +831,38 @@ private static bool CanReplace(ISymbol symbol)
...
+        /// <summary>
+        /// Determines if <paramref name="speculativeSymbol"/> contains a superset of the symbols in
+        /// <paramref name="actualSymbol"/>.
+        /// </summary>
+        private static bool IsReplacementCandidate(SymbolInfo actualSymbol, SymbolInfo speculativeSymbol)
+        {
+            if (!(speculativeSymbol.Symbol is null))
+            {
+                return Equals(actualSymbol.Symbol, speculativeSymbol.Symbol);
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: this block doesn't seem to match the comment for thsi method.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 2]
```diff
@@ -812,6 +831,38 @@ private static bool CanReplace(ISymbol symbol)
...
+
+            if (speculativeSymbol.CandidateSymbols.IsDefault)
+            {
+                return false;
+            }
+
+            if (!(actualSymbol.Symbol is null))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: actualSymbol.Symbol is known to not be null due to the if check above the call to this method, right?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 3]
```diff
@@ -799,6 +799,25 @@ private static bool CanReplace(ISymbol symbol)
...
+            if (!memberAccess.Expression.IsKind(SyntaxKind.ThisExpression, SyntaxKind.BaseExpression))
+            {
+                var actualSymbol = semanticModel.GetSymbolInfo(memberAccess.Name, cancellationToken);
+                if (actualSymbol.GetAnySymbol() is null)
+                {
+                    return false;
+                }
+
+                // Get a syntax node that's not part of the tree to make sure we speculate on just the name.
+                var speculativeTree = memberAccess.Name.WithoutTrivia();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: nit: use 'Node' or 'Name' instead of 'Tree' as Tree implies SyntaxTree.

Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Note: no need to remove trivia.  Which will also save on allocations.

Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Oh, i see, you're removing trivia to get a node that's not in the tree.   Instead of this, could you just do SemanticModel.LookupSymbols with the name you have?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Oh, you call GetAnySymbol up there.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 4]
```diff
@@ -812,6 +831,38 @@ private static bool CanReplace(ISymbol symbol)
             return memberAccess.CanReplaceWithReducedName(replacementNode, semanticModel, cancellationToken);
         }
 
+        /// <summary>
+        /// Determines if <paramref name="speculativeSymbol"/> contains a superset of the symbols in
+        /// <paramref name="actualSymbol"/>.
+        /// </summary>
+        private static bool IsReplacementCandidate(SymbolInfo actualSymbol, SymbolInfo speculativeSymbol)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Overall, shouldn't we only be removing this/base if the name after hte dot binds without errors?  I would think we would only do this if and only if actualSymbl and speculativeSymbol have ".Symbol" set and not if they have Candidates set.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 2 above)
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: ➡️ Yep, I know there exists one or more symbols in `actualSymbol`, but not whether it's in `Symbol` or `CandidateSymbols`.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 3 above)
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Cool, I hadn't used `LookupSymbols` before. It turns out that the complete check needs to use `LookupSymbols` as an initial check (the fastest expected path), and `LookupNamespacesOrTypes` to cover the special case seen in `SimplifyTypeNamesTests.TestColorColorCase1`. This update removed the need to pull `memberAccess.Name` out of the tree and also fixed the previously failing test cases.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 4 above)
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: 1. I believe it's possible that `actualSymbol` contains multiple symbols even in valid code (e.g. `nameof` applied to a method group)
1. The lookup doesn't perform overload resolution for invocations, so it's possible that the speculative symbols to contain multiple possibilities that still resolve to the same symbol in the end

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 5]
```diff
@@ -799,6 +800,36 @@ private static bool CanReplace(ISymbol symbol)
...
+            // 'this.Name' or 'base.Name', no additional check here is required.
+            if (!memberAccess.Expression.IsKind(SyntaxKind.ThisExpression, SyntaxKind.BaseExpression))
+            {
+                var actualSymbol = semanticModel.GetSymbolInfo(memberAccess.Name, cancellationToken);
+                if (actualSymbol.GetAnySymbol() is null)
+                {
+                    return false;
+                }
+
+                var symbols = semanticModel.LookupSymbols(memberAccess.SpanStart, name: memberAccess.Name.Identifier.ValueText);
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: :bulb: Technically we can avoid the call to `LookupSymbols` and go straight for `LookupNamespacesAndTypes` any time `actualSymbol` only contains namespace and/or type symbols. This will actually be a common case for Simplify Type Name.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 1 above)
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: :memo: This code was rewritten after finding bugs in the original implementation.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 5 above)
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: ➡️ This optimization has now been implemented.

**Comment by sharwell** [comment]
@CyrusNajmabadi Any suggestions for handling the remaining edge case? It's conceptually the opposite of what I'm using `LookupNamespacesAndTypes` for.

**Comment by sharwell** [comment]
:memo: So the problem is I don't have a way to specify `LookupOptions.MustBeInvocableIfMember` when calling `SemanticModel.LookupSymbols`. Any suggestions for how to approach this would be appreciated.

**Comment by sharwell** [comment]
Requesting assistance from @gafter or @AlekseyTs for handling the special case. Please let me know if you have any questions; the impact of this change on end users is substantial so I want to make sure I get it right in time to deliver. :smile:

**Comment by Pilchie** [comment]
Pinging @gafter and @AlekseyTs again.  The gist of the issue here is that in this code:
```C#
                using System;
                class A
                {
                    static void Del() { }
                    class B
                    {
                        delegate void Del();
                        void Boo()
                        {
                            Del d = new Del(A.De); 
                            A.Del(); 
                        }
                    }
                }
```

The `A.Del();` line ends up getting not being offered to simplify to just `Del();`.

@sharwell Given the magnitude of this perf win, and the fact that this wouldn't introduce any new diagnostics, I'm also okay with skipping the test and filing a separate issue to follow up.

**Comment by CyrusNajmabadi** [comment]
One thing that could be done would be to just call LookupSymbols at your position.  As long as you keep hitting a type, then walk upwards and check again at the next higher scope.

**Comment by CyrusNajmabadi** [comment]
It's definitely hacky, and it would be great if we could get a LookupSymbols flag for this.  but it seems like something that could at least work in the interim.

**Comment by sharwell** [comment]
@CyrusNajmabadi The hardest part about that is getting the position to pass to `LookupSymbols`.

**Comment by sharwell** [comment]
@MattGertz for ask mode (note the intentional regression/deferral described under Risk)

**Comment by MattGertz** [comment]
Just talked with Kevin & Cyrus and I (strongly) agree with this tradeoff.
